deprecate Task.didWork (#25337)
gradle/gradle: subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc

The embedded Kotlin has been updated to link:https://github.com/JetBrains/kotlin/releases/tag/v1.8.21[Kotlin 1.8.21].

=== Deprecations

[[deprecated_task_did_work]]
==== Deprecated Task.didWork

> The following methods on `Task` are deprecated

Reviewer: ```suggestion
The following methods on `Task` are deprecated:
```

- link:{javadocPath}/org/gradle/api/Task.html#setDidWork-boolean-[setDidWork]
- link:{javadocPath}/org/gradle/api/Task.html#getDidWork--[getDidWork]

> For some tasks, `Task.getDidWork` returns false even if the task did execute but the output

Reviewer: ```suggestion
For some tasks, `Task.getDidWork` returns false even if the task executed and the output(s)
```

> did not change (aka "up-to-date after-the-fact"). This is unexpected, because although the

Reviewer: ```suggestion
did not change (aka "up-to-date after-the-fact"). Although the
```

> task was executed and took time, the output of the task is `UP-TO-DATE`. This will change in

Reviewer: ```suggestion
task executed and took time, the outcome label of the task is `UP-TO-DATE`. 
This behavior is unexpected.
```

> the future, and as a step in that direction, `Task.didWork` was deprecated. You can check if

Reviewer: ```suggestion
As part of ongoing efforts to fix this behavior, `Task.didWork` was deprecated. 
```

> the task was "up-to-date up front" by using

Reviewer: ```suggestion
You can check if a task was "up-to-date up front" by using
```
